homeassistant/components/traccar_server/sensor.py

```diff
@@ -0,0 +1,132 @@
+"""Support for Traccar server sensors."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Generic, Literal, TypeVar, cast
+
+from pytraccar import DeviceModel, GeofenceModel, PositionModel
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import PERCENTAGE, EntityCategory, UnitOfLength, UnitOfSpeed
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+
+from .const import DOMAIN
+from .coordinator import TraccarServerCoordinator
+from .entity import TraccarServerEntity
+
+_T = TypeVar("_T")
+
+
+@dataclass(frozen=True, kw_only=True)
+class TraccarServerSensorEntityDescription(Generic[_T], SensorEntityDescription):
+    """Describe Traccar Server sensor entity."""
+
+    data_key: Literal["position", "device", "geofence", "attributes"]
+    entity_registry_enabled_default = False
+    entity_category = EntityCategory.DIAGNOSTIC
+    value_fn: Callable[[_T], StateType]
+
+
+TRACCAR_SERVER_SENSOR_ENTITY_DESCRIPTIONS = (
+    TraccarServerSensorEntityDescription[PositionModel](
+        key="attributes.batteryLevel",
+        data_key="position",
+        native_unit_of_measurement=PERCENTAGE,
+        device_class=SensorDeviceClass.BATTERY,
+        state_class=SensorStateClass.MEASUREMENT,
+        suggested_display_precision=0,
+        value_fn=lambda x: x["attributes"].get("batteryLevel", -1),
+    ),
+    TraccarServerSensorEntityDescription[PositionModel](
+        key="speed",
+        data_key="position",
+        device_class=SensorDeviceClass.SPEED,
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfSpeed.KNOTS,
+        suggested_display_precision=0,
+        value_fn=lambda x: x["speed"],
+    ),
+    TraccarServerSensorEntityDescription[PositionModel](
+        key="altitude",
+        data_key="position",
+        translation_key="altitude",
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement=UnitOfLength.METERS,
+        suggested_display_precision=1,
+        value_fn=lambda x: x["altitude"],
+    ),
+    TraccarServerSensorEntityDescription[PositionModel](
+        key="address",
+        data_key="position",
+        translation_key="address",
+        value_fn=lambda x: x["address"],
+    ),
+    TraccarServerSensorEntityDescription[DeviceModel](
+        key="status",
+        data_key="device",
+        translation_key="status",
+        device_class=SensorDeviceClass.ENUM,
+        value_fn=lambda x: x["status"],
+        options=["offline", "online", "unknown"],
```

Comment:
As this only has 3 states, this could be a connectivity binary sensor 🤷‍♂️  where "unknown" would be `None`

Reply:
Agree!

Reply:
Excluded the move for the `status` attribute in https://github.com/home-assistant/core/pull/111374/commits/e3b6a9e20b987065a8e8fa8e493add6c68c09b12, will do a followup PR that adds the `binary_sensor` platform to this integration 👍